src/solver_py.cpp

```diff
@@ -49,18 +50,151 @@ void Solver_Py::extract_options(const po::variables_map &vm __attribute__((unuse
 
 }
 
-void Solver_Py::add_edge(const string &src_path __attribute__((unused)),
-                         const string &statement __attribute__((unused)),
-                         unsigned int idx __attribute__((unused)),
-                         unsigned int line_no __attribute__((unused))) {
+void Solver_Py::add_edge(const string &src_path,
+                         const string &statement,
+                         unsigned int idx,
+                         unsigned int line_no) {
+    using boost::filesystem::path;
+    using boost::filesystem::operator/;
+
+    BOOST_LOG_TRIVIAL(trace) << "add_edge: " << src_path << " -> " << statement
+                             << ", idx = " << idx << ", line_no = " << line_no;
+
+    if (statement == "*") return;
+    if (statement.empty()) return;
+
+    // Handle each comma-separated part separately
+    if (contains_string(statement, ",")) {
+        vector<string> comma_separated_statements = separate_string(statement, ',');
+        add_edges(src_path, comma_separated_statements, idx, line_no);
+        return;
+    }
+
+    string module_name = statement;
+
+    if (contains_string(statement, ".")) {
+        module_name = get_final_substring(statement, ".");
+    }
+
+    if (contains_string(statement, " as ")){
+        module_name = get_first_substring(statement, " as ");
+    }
+
+    module_name = remove_whitespaces(module_name);
+
+    unique_lock<mutex> glck(graph_mutex);
```

Comment:
Are we sure that the above is thread safe so that the lock can be done here?

Reply:
I believe that the mutex is there to protect additions and modifications to the graph.  Everything before the lock does not modify the graph and it would be wasteful to hold the lock in this case when only stripping a string. Putting the lock at the start of add_edge() would mean that every thread has to wait to modify their string, when they in this case could already be done and waiting for the lock.